### ISSUE TITLE
2039 onwards date_time_build behaviour

### ISSUE THREAD

**Issue by TimTaylor** [r-lib/clock#292]
Something seems to change in behaviour with `date_time_build()` in 2039 and how it's handling daylight saving times. I may be missing something but I'm a little confused. Dev version used below but same behaviour in released:
``` r
library(clock)
packageVersion("clock")
#> [1] '0.6.0.9000'
.mapply(
    date_time_build,
    dots = list(year=2031:2040),
    MoreArgs = list(month=4, day=1:10, zone = "NZ")
)
#> [[1]]
#>  [1] "2031-04-01 NZDT" "2031-04-02 NZDT" "2031-04-03 NZDT" "2031-04-04 NZDT"
#>  [5] "2031-04-05 NZDT" "2031-04-06 NZDT" "2031-04-07 NZST" "2031-04-08 NZST"
#>  [9] "2031-04-09 NZST" "2031-04-10 NZST"
#> 
#> [[2]]
#>  [1] "2032-04-01 NZDT" "2032-04-02 NZDT" "2032-04-03 NZDT" "2032-04-04 NZDT"
#>  [5] "2032-04-05 NZST" "2032-04-06 NZST" "2032-04-07 NZST" "2032-04-08 NZST"
#>  [9] "2032-04-09 NZST" "2032-04-10 NZST"
#> 
#> [[3]]
#>  [1] "2033-04-01 NZDT" "2033-04-02 NZDT" "2033-04-03 NZDT" "2033-04-04 NZST"
#>  [5] "2033-04-05 NZST" "2033-04-06 NZST" "2033-04-07 NZST" "2033-04-08 NZST"
#>  [9] "2033-04-09 NZST" "2033-04-10 NZST"
#> 
#> [[4]]
#>  [1] "2034-04-01 NZDT" "2034-04-02 NZDT" "2034-04-03 NZST" "2034-04-04 NZST"
#>  [5] "2034-04-05 NZST" "2034-04-06 NZST" "2034-04-07 NZST" "2034-04-08 NZST"
#>  [9] "2034-04-09 NZST" "2034-04-10 NZST"
#> 
#> [[5]]
#>  [1] "2035-04-01 NZDT" "2035-04-02 NZST" "2035-04-03 NZST" "2035-04-04 NZST"
#>  [5] "2035-04-05 NZST" "2035-04-06 NZST" "2035-04-07 NZST" "2035-04-08 NZST"
#>  [9] "2035-04-09 NZST" "2035-04-10 NZST"
#> 
#> [[6]]
#>  [1] "2036-04-01 NZDT" "2036-04-02 NZDT" "2036-04-03 NZDT" "2036-04-04 NZDT"
#>  [5] "2036-04-05 NZDT" "2036-04-06 NZDT" "2036-04-07 NZST" "2036-04-08 NZST"
#>  [9] "2036-04-09 NZST" "2036-04-10 NZST"
#> 
#> [[7]]
#>  [1] "2037-04-01 NZDT" "2037-04-02 NZDT" "2037-04-03 NZDT" "2037-04-04 NZDT"
#>  [5] "2037-04-05 NZDT" "2037-04-06 NZST" "2037-04-07 NZST" "2037-04-08 NZST"
#>  [9] "2037-04-09 NZST" "2037-04-10 NZST"
#> 
#> [[8]]
#>  [1] "2038-04-01 NZDT" "2038-04-02 NZDT" "2038-04-03 NZDT" "2038-04-04 NZDT"
#>  [5] "2038-04-05 NZST" "2038-04-06 NZST" "2038-04-07 NZST" "2038-04-08 NZST"
#>  [9] "2038-04-09 NZST" "2038-04-10 NZST"
#> 
#> [[9]]
#>  [1] "2039-04-01 00:00:00 NZDT" "2039-04-02 00:00:00 NZDT"
#>  [3] "2039-04-03 00:00:00 NZDT" "2039-04-04 01:00:00 NZDT"
#>  [5] "2039-04-05 01:00:00 NZDT" "2039-04-06 01:00:00 NZDT"
#>  [7] "2039-04-07 01:00:00 NZDT" "2039-04-08 01:00:00 NZDT"
#>  [9] "2039-04-09 01:00:00 NZDT" "2039-04-10 01:00:00 NZDT"
#> 
#> [[10]]
#>  [1] "2040-04-01 00:00:00 NZDT" "2040-04-02 01:00:00 NZDT"
#>  [3] "2040-04-03 01:00:00 NZDT" "2040-04-04 01:00:00 NZDT"
#>  [5] "2040-04-05 01:00:00 NZDT" "2040-04-06 01:00:00 NZDT"
#>  [7] "2040-04-07 01:00:00 NZDT" "2040-04-08 01:00:00 NZDT"
#>  [9] "2040-04-09 01:00:00 NZDT" "2040-04-10 01:00:00 NZDT"
```

<sup>Created on 2022-02-28 by the [reprex package](https://reprex.tidyverse.org) (v2.0.1)</sup>

**Comment by TimTaylor** [comment]
Apologies - this is R not clock

**Comment by DavisVaughan** [comment]
Did you determine that it is a printing issue on Linux?

Can you try this code? Do you get the "right" result when printing out the zoned-time on Linux, but not with the POSIXct?

``` r
library(clock)
library(magrittr)

dates <- .mapply(
  year_month_day,
  dots = list(year=2031:2040),
  MoreArgs = list(month=4, day=1:10)
) %>%
  lapply(as_naive_time) %>%
  lapply(as_zoned_time, zone = "NZ")

# Hopefully this looks right
dates
#> [[1]]
#> <zoned_time<second><NZ>[10]>
#>  [1] "2031-04-01T00:00:00+13:00" "2031-04-02T00:00:00+13:00"
#>  [3] "2031-04-03T00:00:00+13:00" "2031-04-04T00:00:00+13:00"
#>  [5] "2031-04-05T00:00:00+13:00" "2031-04-06T00:00:00+13:00"
#>  [7] "2031-04-07T00:00:00+12:00" "2031-04-08T00:00:00+12:00"
#>  [9] "2031-04-09T00:00:00+12:00" "2031-04-10T00:00:00+12:00"
#> 
#> [[2]]
#> <zoned_time<second><NZ>[10]>
#>  [1] "2032-04-01T00:00:00+13:00" "2032-04-02T00:00:00+13:00"
#>  [3] "2032-04-03T00:00:00+13:00" "2032-04-04T00:00:00+13:00"
#>  [5] "2032-04-05T00:00:00+12:00" "2032-04-06T00:00:00+12:00"
#>  [7] "2032-04-07T00:00:00+12:00" "2032-04-08T00:00:00+12:00"
#>  [9] "2032-04-09T00:00:00+12:00" "2032-04-10T00:00:00+12:00"
#> 
#> [[3]]
#> <zoned_time<second><NZ>[10]>
#>  [1] "2033-04-01T00:00:00+13:00" "2033-04-02T00:00:00+13:00"
#>  [3] "2033-04-03T00:00:00+13:00" "2033-04-04T00:00:00+12:00"
#>  [5] "2033-04-05T00:00:00+12:00" "2033-04-06T00:00:00+12:00"
#>  [7] "2033-04-07T00:00:00+12:00" "2033-04-08T00:00:00+12:00"
#>  [9] "2033-04-09T00:00:00+12:00" "2033-04-10T00:00:00+12:00"
#> 
#> [[4]]
#> <zoned_time<second><NZ>[10]>
#>  [1] "2034-04-01T00:00:00+13:00" "2034-04-02T00:00:00+13:00"
#>  [3] "2034-04-03T00:00:00+12:00" "2034-04-04T00:00:00+12:00"
#>  [5] "2034-04-05T00:00:00+12:00" "2034-04-06T00:00:00+12:00"
#>  [7] "2034-04-07T00:00:00+12:00" "2034-04-08T00:00:00+12:00"
#>  [9] "2034-04-09T00:00:00+12:00" "2034-04-10T00:00:00+12:00"
#> 
#> [[5]]
#> <zoned_time<second><NZ>[10]>
#>  [1] "2035-04-01T00:00:00+13:00" "2035-04-02T00:00:00+12:00"
#>  [3] "2035-04-03T00:00:00+12:00" "2035-04-04T00:00:00+12:00"
#>  [5] "2035-04-05T00:00:00+12:00" "2035-04-06T00:00:00+12:00"
#>  [7] "2035-04-07T00:00:00+12:00" "2035-04-08T00:00:00+12:00"
#>  [9] "2035-04-09T00:00:00+12:00" "2035-04-10T00:00:00+12:00"
#> 
#> [[6]]
#> <zoned_time<second><NZ>[10]>
#>  [1] "2036-04-01T00:00:00+13:00" "2036-04-02T00:00:00+13:00"
#>  [3] "2036-04-03T00:00:00+13:00" "2036-04-04T00:00:00+13:00"
#>  [5] "2036-04-05T00:00:00+13:00" "2036-04-06T00:00:00+13:00"
#>  [7] "2036-04-07T00:00:00+12:00" "2036-04-08T00:00:00+12:00"
#>  [9] "2036-04-09T00:00:00+12:00" "2036-04-10T00:00:00+12:00"
#> 
#> [[7]]
#> <zoned_time<second><NZ>[10]>
#>  [1] "2037-04-01T00:00:00+13:00" "2037-04-02T00:00:00+13:00"
#>  [3] "2037-04-03T00:00:00+13:00" "2037-04-04T00:00:00+13:00"
#>  [5] "2037-04-05T00:00:00+13:00" "2037-04-06T00:00:00+12:00"
#>  [7] "2037-04-07T00:00:00+12:00" "2037-04-08T00:00:00+12:00"
#>  [9] "2037-04-09T00:00:00+12:00" "2037-04-10T00:00:00+12:00"
#> 
#> [[8]]
#> <zoned_time<second><NZ>[10]>
#>  [1] "2038-04-01T00:00:00+13:00" "2038-04-02T00:00:00+13:00"
#>  [3] "2038-04-03T00:00:00+13:00" "2038-04-04T00:00:00+13:00"
#>  [5] "2038-04-05T00:00:00+12:00" "2038-04-06T00:00:00+12:00"
#>  [7] "2038-04-07T00:00:00+12:00" "2038-04-08T00:00:00+12:00"
#>  [9] "2038-04-09T00:00:00+12:00" "2038-04-10T00:00:00+12:00"
#> 
#> [[9]]
#> <zoned_time<second><NZ>[10]>
#>  [1] "2039-04-01T00:00:00+13:00" "2039-04-02T00:00:00+13:00"
#>  [3] "2039-04-03T00:00:00+13:00" "2039-04-04T00:00:00+12:00"
#>  [5] "2039-04-05T00:00:00+12:00" "2039-04-06T00:00:00+12:00"
#>  [7] "2039-04-07T00:00:00+12:00" "2039-04-08T00:00:00+12:00"
#>  [9] "2039-04-09T00:00:00+12:00" "2039-04-10T00:00:00+12:00"
#> 
#> [[10]]
#> <zoned_time<second><NZ>[10]>
#>  [1] "2040-04-01T00:00:00+13:00" "2040-04-02T00:00:00+12:00"
#>  [3] "2040-04-03T00:00:00+12:00" "2040-04-04T00:00:00+12:00"
#>  [5] "2040-04-05T00:00:00+12:00" "2040-04-06T00:00:00+12:00"
#>  [7] "2040-04-07T00:00:00+12:00" "2040-04-08T00:00:00+12:00"
#>  [9] "2040-04-09T00:00:00+12:00" "2040-04-10T00:00:00+12:00"

# This could look wrong on Ubuntu? I guess?
lapply(dates, as.POSIXct)
#> [[1]]
#>  [1] "2031-04-01 NZDT" "2031-04-02 NZDT" "2031-04-03 NZDT" "2031-04-04 NZDT"
#>  [5] "2031-04-05 NZDT" "2031-04-06 NZDT" "2031-04-07 NZST" "2031-04-08 NZST"
#>  [9] "2031-04-09 NZST" "2031-04-10 NZST"
#> 
#> [[2]]
#>  [1] "2032-04-01 NZDT" "2032-04-02 NZDT" "2032-04-03 NZDT" "2032-04-04 NZDT"
#>  [5] "2032-04-05 NZST" "2032-04-06 NZST" "2032-04-07 NZST" "2032-04-08 NZST"
#>  [9] "2032-04-09 NZST" "2032-04-10 NZST"
#> 
#> [[3]]
#>  [1] "2033-04-01 NZDT" "2033-04-02 NZDT" "2033-04-03 NZDT" "2033-04-04 NZST"
#>  [5] "2033-04-05 NZST" "2033-04-06 NZST" "2033-04-07 NZST" "2033-04-08 NZST"
#>  [9] "2033-04-09 NZST" "2033-04-10 NZST"
#> 
#> [[4]]
#>  [1] "2034-04-01 NZDT" "2034-04-02 NZDT" "2034-04-03 NZST" "2034-04-04 NZST"
#>  [5] "2034-04-05 NZST" "2034-04-06 NZST" "2034-04-07 NZST" "2034-04-08 NZST"
#>  [9] "2034-04-09 NZST" "2034-04-10 NZST"
#> 
#> [[5]]
#>  [1] "2035-04-01 NZDT" "2035-04-02 NZST" "2035-04-03 NZST" "2035-04-04 NZST"
#>  [5] "2035-04-05 NZST" "2035-04-06 NZST" "2035-04-07 NZST" "2035-04-08 NZST"
#>  [9] "2035-04-09 NZST" "2035-04-10 NZST"
#> 
#> [[6]]
#>  [1] "2036-04-01 NZDT" "2036-04-02 NZDT" "2036-04-03 NZDT" "2036-04-04 NZDT"
#>  [5] "2036-04-05 NZDT" "2036-04-06 NZDT" "2036-04-07 NZST" "2036-04-08 NZST"
#>  [9] "2036-04-09 NZST" "2036-04-10 NZST"
#> 
#> [[7]]
#>  [1] "2037-04-01 NZDT" "2037-04-02 NZDT" "2037-04-03 NZDT" "2037-04-04 NZDT"
#>  [5] "2037-04-05 NZDT" "2037-04-06 NZST" "2037-04-07 NZST" "2037-04-08 NZST"
#>  [9] "2037-04-09 NZST" "2037-04-10 NZST"
#> 
#> [[8]]
#>  [1] "2038-04-01 NZDT" "2038-04-02 NZDT" "2038-04-03 NZDT" "2038-04-04 NZDT"
#>  [5] "2038-04-05 NZST" "2038-04-06 NZST" "2038-04-07 NZST" "2038-04-08 NZST"
#>  [9] "2038-04-09 NZST" "2038-04-10 NZST"
#> 
#> [[9]]
#>  [1] "2039-04-01 NZDT" "2039-04-02 NZDT" "2039-04-03 NZDT" "2039-04-04 NZST"
#>  [5] "2039-04-05 NZST" "2039-04-06 NZST" "2039-04-07 NZST" "2039-04-08 NZST"
#>  [9] "2039-04-09 NZST" "2039-04-10 NZST"
#> 
#> [[10]]
#>  [1] "2040-04-01 NZDT" "2040-04-02 NZST" "2040-04-03 NZST" "2040-04-04 NZST"
#>  [5] "2040-04-05 NZST" "2040-04-06 NZST" "2040-04-07 NZST" "2040-04-08 NZST"
#>  [9] "2040-04-09 NZST" "2040-04-10 NZST"
```

<sup>Created on 2022-02-28 by the [reprex package](https://reprex.tidyverse.org) (v2.0.1)</sup>

**Comment by TimTaylor** [comment]
@DavisVaughan - yes to confirm:
- clock's zoned time prints without issue.
- second example prints differently post 2038 due to something with Ubuntu 18.04 and/or the R build I'm using for 18.04. Not {clock} related and fixed with 20.04 or other more recent distros so assume it is something that has been patched.

Sorry for the noise.
